### PR TITLE
Add shuffle button to movie collections

### PR DESCRIPTION
**Pull request by bernarden**
**Changes**
Currently shuffling movies (tv shows and music as well) is possible only on desktop by right clicking on collection and then "Shuffle", so mobile apps and tv apps are missing out on this feature. This PR adds a shuffle button for movie collections.
image

**Issues**
This issue from 2020 is the closest that I could find but for music: #1834
Side note: Stale bot should not be closing issues that have been manually tagged as bugs.

### CODE REVIEW

File: src/controllers/movies/movies.html [hunk 1]
```diff
@@ -6,6 +6,7 @@
             <button is="paper-icon-button-light" class="btnSelectView autoSize" title="${ButtonSelectView}"><span class="material-icons view_comfy" aria-hidden="true"></span></button>
             <button is="paper-icon-button-light" class="btnSort autoSize" title="${Sort}"><span class="material-icons sort_by_alpha" aria-hidden="true"></span></button>
             <button is="paper-icon-button-light" class="btnFilter autoSize" title="${Filter}"><span class="material-icons filter_list" aria-hidden="true"></span></button>
+            <button is="paper-icon-button-light" class="btnShuffle autoSize" title="${Shuffle}"><span class="material-icons shuffle" aria-hidden="true"></span></button>
```
Comment (dmitrylyzo) [inline src/controllers/movies/movies.html:9]: Maybe move it before `btnSelectView` - to follow MusicAlbums' order?

Comment (bernarden) [reply src/controllers/movies/movies.html]: Resolved.

File: src/controllers/movies/movies.js [hunk 2]
```diff
@@ -77,6 +91,10 @@ import '../../elements/emby-itemscontainer/emby-itemscontainer';
                 elem.addEventListener('click', onPreviousPageClick);
             }
 
+            for (const elem of tabContent.querySelectorAll('.btnShuffle')) {
+                elem.addEventListener('click', onShuffleClick);
+            }
+
```
[the remark this answers is not in the document]
Comment (bernarden) [reply src/controllers/movies/movies.js]: Resolved.

**Comment by dmitrylyzo** [comment]
We need to hide the `btnShuffle` button in `afterRefresh` if the library is empty (before autoFocus).
_Currently, we are getting an endless "loading"._

I haven't tested with broken items. It probably just skips them, but I don't remember.

**Comment by bernarden** [comment]
> re: stale bot we have a `confirmed` label for issues that have been replicated that prevents the automatic closing.
> 
> I feel our current configuration for automatic closing is really quiet generous, after ~4 months (120 days) with no activity on an issue a warning is provided and it is labeled as stale, then if another 3 weeks passes with no activity, the issue is closed. We have very few people triaging tickets, so stale bot is a necessity for us to keep a relevant backlog of issues.

My statement was based on what I saw in that one issue and to me it looked like the issue was raised and marked as a bug then no one had time get to it in 4 months and it stopped being an issue. I guess it just needed a `confirmed` label then.

**Comment by bernarden** [comment]
> We need to hide the `btnShuffle` button in `afterRefresh` if the library is empty (before autoFocus). _Currently, we are getting an endless "loading"._
> 
> I haven't tested with broken items. It probably just skips them, but I don't remember.

Good find. I hid the shuffle button when there are no items displayed but this is not the best fix since shuffle button functionality currently shuffles the whole collection (not impacted by filters) but the shuffle button element can still get hidden when certain filters are applied.

File: src/controllers/movies/movies.js [hunk 3]
```diff
@@ -77,6 +87,8 @@ import '../../elements/emby-itemscontainer/emby-itemscontainer';
                 elem.addEventListener('click', onPreviousPageClick);
             }
 
+            tabContent.querySelector(".btnShuffle").classList.toggle("hide", result.TotalRecordCount < 1);
```
Comment (dmitrylyzo) [inline src/controllers/movies/movies.js:90]:
```suggestion
            tabContent.querySelector('.btnShuffle').classList.toggle('hide', result.TotalRecordCount < 1);
```

**Review by dmitrylyzo — COMMENTED** [review]
Yeah, filtering doesn't apply in any view, iirc.
Obviously, we need to fix that endless "loading" (not in this PR).

**Review by dmitrylyzo — APPROVED** [review]
LGTM

**Comment by ireun** [comment]
Is this feature broken or something? I'm not getting this on my movies library.

**Comment by dmitrylyzo** [comment]
> Is this feature broken or something? I'm not getting this on my movies library.

The repo was in a feature freeze (pre 10.8), so it will only be released in version 10.9

**Comment by nonobio** [comment]
Hi,
Is there a workaround by waiting the 10.9 version?
Thanks

**Pull request merged**